R/dataframe-s3-base.R

```diff
@@ -121,3 +121,209 @@ head.polars_data_frame <- function(x, n = 6L, ...) x$head(n = n)
 
 #' @exportS3Method utils::tail
 tail.polars_data_frame <- function(x, n = 6L, ...) x$tail(n = n)
+
+#' Try to match `tibble` behavior as much as possible, following
+#' https://tibble.tidyverse.org/articles/invariants.html#column-subsetting
+#' @export
+#' @noRd
+`[.polars_data_frame` <- function(x, i, j, drop = FALSE) {
+  cols <- names(x)
+
+  # useful for error messages below
+  i_arg <- substitute(i)
+  j_arg <- substitute(j)
+
+  # taken from tibble:::`[`
+  if (missing(i)) {
+    i <- NULL
+    i_arg <- NULL
+  } else if (is.null(i)) {
+    i <- integer()
+  }
+  if (missing(j)) {
+    j <- NULL
+    j_arg <- NULL
+  } else if (is.null(j)) {
+    j <- integer()
+  }
+  n_real_args <- nargs() - !missing(drop)
+  if (n_real_args <= 2L) {
+    if (!missing(drop)) {
+      warn(
+        c(
+          "!" = "`drop` argument ignored for subsetting a frame with `x[j]`.",
+          "i" = "It has an effect only for `x[i, j]`."
+        )
+      )
+      drop <- FALSE
+    }
+    j <- i
+    i <- NULL
+    j_arg <- i_arg
+    i_arg <- NULL
+  }
+
+  if (is.null(i) && is.null(j)) {
+    return(x)
+  }
+
+  #### Rows -----------------------------------------------------
+
+  # check accepted types for subsetting rows
+  if (!is.null(i) && !is.character(i) && !is.numeric(i) && !is.logical(i)) {
+    abort(
+      c(
+        sprintf("Can't subset rows with `%s`.", deparse(i_arg)),
+        "i" = sprintf(
+          "`%s` must be logical, numeric, or character, not %s.",
+          deparse(i_arg),
+          obj_type_friendly(i)
+        )
+      )
+    )
+  }
+
+  if (!is.null(i)) {
+    # If logical, `i` must be of length 1 or number of rows
+    if (is.logical(i)) {
+      if (length(i) == 1 && isTRUE(i)) {
+        idx <- rep_len(TRUE, x$height)
+      } else if (length(i) == x$height) {
+        idx <- i
+      } else {
+        abort(
+          c(
+            "!" = sprintf("Can't subset rows with `%s`.", deparse(i_arg)),
+            "i" = sprintf(
+              "Logical subscript `%s` must be size 1 or %s, not %s",
+              deparse(i_arg),
+              x$height,
+              length(i)
+            )
+          )
+        )
+      }
+    } else {
+      # Negative indices -> drop those rows
+      # Do not accept mixing negative and positive indices
+      if (all(i < 0)) {
```

Comment:
Is this operation safe? (i.e., does `i < 0` always return a boolean value, or can it be `NA` or non-numeric?)